proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,507 @@
+to evolve since.
+
+# Dynamic Supervisor OIDC Clients
+
+## Problem Statement
+
+Pinniped can be used to provide authentication to Kubernetes clusters via `kubectl` for cluster users such as
+developers, devops teams, and cluster admins. However, sometimes these same users need to be able to authenticate to
+webapps running on these clusters to perform actions such as installing, configuring, and monitoring applications on the
+cluster. It would be fitting for Pinniped to also provide authentication for these types of webapps, to ensure that the
+same users can authenticate in exactly the same way, using the same identity provider, and resolving their identities to
+the same usernames and group memberships. Enabling this use case will require new features in Pinniped, which are
+proposed in this document.
+
+### How Pinniped Works Today (as of version v0.15.0)
+
+Each
+[FederationDomain](https://github.com/vmware-tanzu/pinniped/blob/main/generated/1.23/README.adoc#federationdomain)
+configured in the Pinniped Supervisor is an OIDC Provider issuer which implements
+the [OIDC authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).
+
+Today, the Pinniped Supervisor only allows one hardcoded OIDC client, called `pinniped-cli`. This client is only allowed
+to redirect authcodes to the CLI's localhost listener. This makes it intentionally impossible for this client to be used
+by a webapp running on the cluster (or anywhere else on the network). The `pinniped-cli` client is implicitly available
+on all FederationDomains configured in the Supervisor, since every FederationDomain allows users to authenticate
+themselves via the Pinniped CLI for `kubectl` integration.
+
+## Terminology / Concepts
+
+- See the [definition of a "client" in the OAuth 2.0 spec](https://datatracker.ietf.org/doc/html/rfc6749#section-1.1).
+  For the purposes of this proposal, a "client" is roughly equal to a webapp which wants to know the authenticated
+  identity of a user, and may want to perform actions as that user on clusters. An admin needs to allow the client to
+  learn about the identity of the users by registering the client with the Pinniped Supervisor.
+- See also the [OIDC terminology in the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#Terminology).
+- The OIDC clients proposed in this document are "dynamic" in the sense that they can be configured and reconfigured on
+  a running Supervisor by the admin.
+
+## Proposal
+
+### Goals and Non-goals
+
+Goals for this proposal:
+
+- Allow Pinniped admins to configure applications (OIDC clients) other than the Pinniped CLI to interact with the
+  Supervisor.
+- Provide a mechanism which governs a client's access to the token exchange APIs. Not all webapps should have permission
+  to act on behalf of the user with the Kubernetes API of the clusters, so an admin must be able to configure which
+  clients have this permission.
```

Comment:
@absoludity Sorry for not being more clear here. There are two token exchange APIs.

The first is the OIDC token endpoint of the Supervisor. It can be used to exchange an OIDC access token for a new cluster-scoped OIDC ID token (with the audience changed to be the audience name of the workload cluster). That's the exchange to which this bullet point refers. An OIDC may or may not be given permission to perform this kind of exchange.

The other is the TokenCredentialRequest of the Concierge that you mention. That endpoint is available to anon users as you pointed out. This proposal does not suggest any changes to that API.

Reply:
Thanks - right, I've not ever used supervisor as we weren't able to provide a different client (other issue, which may be fixed now), so much of my commenting here may have been in the wrong context, sorry.